core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1783,8 +1788,8 @@ protected Scope visitTable(Table table, Optional<Scope> scope)
 
             Optional<MaterializedViewDefinition> optionalMaterializedView = metadata.getMaterializedView(session, name);
             if (optionalMaterializedView.isPresent()) {
-                if (metadata.getMaterializedViewFreshness(session, name).isMaterializedViewFresh()) {
-                    // If materialized view is current, answer the query using the storage table
+                if (isMaterializedViewSufficientlyFresh(name)) {
```

Comment:
I'm wondering if we should let this be an implementation detail of the the connector's ``metadata.getMaterializedViewFreshness` code. Iceberg MV code could figure out the "sufficiently fresh" condition and indicate freshness to the engine through existing boolean flag.

Reply:
We need to differentiate two use-cases

- staleness for explicit querying
  - i didn't yet grasp why we need this, but Trino MVs accept no staleness today, unlike eg PostgreSQL's or Oracle's.  In those systems, the user has control over when the view is refreshed (as in Trino), so it feels OK to query the MV directly if the user asks to, irrespective of the staleness. We chose hybrid approach, but -- for performance and cost control reasons -- users want to query stale MVs directly.  
     i think it's fine to have this configurable on per view basis, but I expect we will later need to add session-level controls (overrides) as well.
    The question is how to we model this for per-MV. I can easily make this a connector property, but this is something we will want to define for all connectors, which calls for some unification (syntax?).
    If we agree that we want to have both per-MV and per-session controls, does it matter which one we start with?
- staleness for query rewrites
   - this one is more tricky. I believe our MVs' "fresh or inline" philosophy was picked with query rewrites as the primary use-case (so departing from apparent experience of other systems). The query rewrites will also want to accept certain staleness, but this use-case is much more delicate, as the logic is implicit. I am sure we will want per-MV and per-session controls for this feature when we build it (but building it is not a goal for this PR)